README.md
# create-favicon

[![npm version](https://img.shields.io/npm/v/create-favicon.svg?style=flat-square)](https://www.npmjs.com/package/create-favicon)

Creates a minimal set of favicons that are compatible with most environments,
based on [this article](https://evilmartians.com/chronicles/how-to-favicon-in-2021-six-files-that-fit-most-needs) by Evil Martians.

Source image must be at least 512x512, preferably a square SVG.

The generated files are:

- 32x32 favicon (for older devices)
- 180x180 PNG icon for Apple devices
- 192x192 PNG icon for Android devices
- 512x512 PNG icon for Android devices
- Web manifest files pointing to the two larger PNG icons
- (If source is an SVG) An SVG icon for modern browsers - this also supports dark mode

Outputs the HTML needed to link these files up on success.
Available as a CLI tool and as an API.

## CLI usage

```sh
## Install globally and run
npm install -g create-favicon
create-favicon <path-to-image> [output-dir]

## Run straight from npm (separate flags with --)
npm create favicon <path-to-image> [output-dir]
npm create favicon <path-to-image> [output-dir] -- --overwrite

## Or use npx (separate flags with --)
npx create-favicon <path-to-image> [output-dir]
npx create-favicon <path-to-image> [output-dir] -- --no-warn
```

## CLI options

```
<path-to-image>     Source image
[output-dir]        Output directory (default: ./favicons)

--overwrite         Overwrite existing files (default: false)
--base-path <path>  Base path for printed HTML and web manifest (default: /)
--no-warn           Disable warnings (default: false)
--no-manifest       Skip outputting a webmanifest (default: false)
-h, --help          Display this message
-v, --version       Display version number
```

## API usage

```ts
import {createFavicon} from 'create-favicon'

// Outputs the files to the current working directory + /favicons,
// and returns the HTML to be inserted into the <head> of your HTML document
const {html} = await createFavicon({sourceFile: '/path/to/some/file.svg'})

// You can also specify the output directory:
const {html} = await createFavicon({
  sourceFile: '/path/to/some/file.svg',
  outputDir: '/path/to/output/dir',
})

// If your files are not going to be placed at the root of your domain,
// you will have to specify a custom base path:
const {html} = await createFavicon({
  sourceFile: '/path/to/some/file.svg',
  basePath: '/my/app',
})

// If you want to overwrite existing files, pass the `overwrite` option:
const {html} = await createFavicon({
  sourceFile: '/path/to/some/file.svg',
  overwrite: true,
})

// Warnings can be silenced by passing `false` to the `warn` option:
const {html} = await createFavicon({
  sourceFile: '/path/to/some/file.svg',
  warn: false,
})

// Manifest generation can be disabled:
const {html} = await createFavicon({
  sourceFile: '/path/to/some/file.svg',
  manifest: false,
})

// The favicon generator can also take a URL as input:
const {html} = await createFavicon({sourceFile: 'https://example.com/sourceLogo.svg'})

// For advanced HTTP cases (need to configure proxies, redirects etc),
// pull the image down yourself and pass it to the generator as a buffer:
const myImage = await someFetcher({
  url: 'https://example.com/sourceLogo.svg',
}).then((res) => res.buffer())

const {html} = await createFavicon({sourceFile: myImage})
```

## License

MIT © [Espen Hovlandsdal](https://espen.codes/)


CHANGELOG.md
<!-- markdownlint-disable --><!-- textlint-disable -->

# 📓 Changelog

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.1.0](https://github.com/rexxars/create-favicon/compare/v2.0.0...v2.1.0) (2023-04-18)

### Features

- use consistent filenames (icon => favicon) ([f911176](https://github.com/rexxars/create-favicon/commit/f911176c552c3f0636052e774c4a53ca880756f0))

## [2.0.0](https://github.com/rexxars/create-favicon/compare/v1.1.0...v2.0.0) (2023-04-18)

### ⚠ BREAKING CHANGES

- `--output-dir` is no longer a valid flag for CLI command,
  instead pass it as the second positional argument. This eases use with
  `npm create` and similar, since you no longer need args separator (`--`) for the
  basic use case.

### Features

- allow disabling web manifest creation ([a6bba03](https://github.com/rexxars/create-favicon/commit/a6bba03cad9dd997146ce694fb26a84c8fe0aea0))

### Code Refactoring

- make output-dir a positional argument instead of flag ([ea2e1fb](https://github.com/rexxars/create-favicon/commit/ea2e1fb0c5d45738771725fd635f01135392e959))

## [1.1.0](https://github.com/rexxars/create-favicon/compare/v1.0.0...v1.1.0) (2023-04-17)

### Features

- do not overwrite files by default, add `--overwrite` flag ([63ec230](https://github.com/rexxars/create-favicon/commit/63ec23022e5909f34b9cf103fac21e06054e9613))


test/api.test.ts
import {createReadStream, existsSync} from 'node:fs'
import {stat, readFile, mkdir, rm, copyFile} from 'node:fs/promises'
import {join as joinPath} from 'node:path'
import {tmpdir} from 'node:os'
import {createHash} from 'node:crypto'
import {type Server, createServer} from 'node:http'
import {afterAll, beforeAll, describe, expect, test, vi} from 'vitest'
import sharp from 'sharp'
import {createFavicon} from '../src'

let tmpDirNum = 0
const testOutputDir = joinPath(tmpdir(), 'favicons-api-output')
const fixturesPath = joinPath(__dirname, 'fixtures')

function getTmpDir(stub = ''): string {
  return joinPath(testOutputDir, `favicon-${stub ? `${stub}-` : ''}${++tmpDirNum}`)
}

async function hashFile(filePath: string): Promise<string> {
  return createHash('sha256')
    .update(await readFile(filePath))
    .digest('hex')
}

describe('api', () => {
  let server: Server

  beforeAll(async () => {
    await mkdir(testOutputDir, {recursive: true})
    await new Promise<void>((resolve, reject) => {
      server = createServer((req, res) => {
        if (req.url === '/mead.svg') {
          res.writeHead(200, 'OK', {'Content-Type': 'image/svg+xml'})
          createReadStream(joinPath(fixturesPath, 'mead.svg')).pipe(res)
          return
        }

        res.writeHead(404, 'Not Found', {'Content-Type': 'text/plain'}).write('Not Found')
        res.end()
      })

      server.on('error', reject)
      server.on('listening', resolve)
      server.listen(27344)
    })
  })

  afterAll(async () => {
    await rm(testOutputDir, {recursive: true})
    await new Promise<void>((resolve, reject) => {
      server.close((err) => (err ? reject(err) : resolve()))
    })
  })

  test('should throw an error if no options are specified', async () => {
    // @ts-expect-error -- We're testing the error case
    await expect(() => createFavicon()).rejects.toMatchInlineSnapshot(
      '[Error: No options specified]'
    )
  })

  test('should throw an error if no source file is specified', async () => {
    // @ts-expect-error -- We're testing the error case
    await expect(() => createFavicon({outputDir: getTmpDir()})).rejects.toMatchInlineSnapshot(
      '[Error: No source file specified]'
    )
  })

  test('should throw an error if the source file is not a string, Buffer, or URL (number input)', async () => {
    await expect(() =>
      // @ts-expect-error -- We're testing the error case
      createFavicon({sourceFile: 123, outputDir: getTmpDir()})
    ).rejects.toMatchInlineSnapshot(
      '[Error: Source file must be a string (file path or URL) or a Buffer]'
    )
  })

  test('should throw an error if the source file is not a string, Buffer, or URL (object input)', async () => {
    await expect(() =>
      // @ts-expect-error -- We're testing the error case
      createFavicon({sourceFile: {foo: 'bar'}, outputDir: getTmpDir()})
    ).rejects.toMatchInlineSnapshot(
      '[Error: Source file must be a string (file path or URL) or a Buffer]'
    )
  })

  test('should throw if file path does not exist', async () => {
    await expect(() =>
      createFavicon({sourceFile: 'does-not-exist.png', outputDir: getTmpDir()})
    ).rejects.toThrow(/does-not-exist\.png": ENOENT/)
  })

  test('should throw if url does not return HTTP 200', async () => {
    await expect(() =>
      createFavicon({sourceFile: 'http://localhost:27344', outputDir: getTmpDir()})
    ).rejects.toThrow(
      'Failed fetching image from "http://localhost:27344": Server returned HTTP 404'
    )
  })

  test('should throw if file could not be parsed as image (non-image format)', async () => {
    await expect(() =>
      createFavicon({sourceFile: joinPath(fixturesPath, 'nonImage.svg'), outputDir: getTmpDir()})
    ).rejects.toThrow(/input buffer contains unsupported image format/i)
  })

  test('should throw if file could not be parsed as image (empty buffer)', async () => {
    await expect(() =>
      createFavicon({sourceFile: Buffer.from(''), outputDir: getTmpDir()})
    ).rejects.toThrow(/input buffer is empty/i)
  })

  test('should throw on too small images', async () => {
    await expect(() =>
      createFavicon({sourceFile: joinPath(fixturesPath, 'tooSmall.png'), outputDir: getTmpDir()})
    ).rejects.toThrow(/source image must be at least 512x512 pixels/i)
  })

  test('should warn on non-square images (console warner)', async () => {
    const spy = vi.spyOn(console, 'warn').mockImplementation(() => null)
    await createFavicon({
      sourceFile: joinPath(fixturesPath, 'nonSquare.svg'),
      outputDir: getTmpDir(),
    })
    expect(spy).toHaveBeenCalledWith(
      '\u001b[93m[warn]\u001b[39m Source image is not square - it is HIGHLY recommended that input image is square'
    )
    spy.mockRestore()
  })

  test('should warn on non-square images (custom warner)', async () => {
    const warn = vi.fn()
    await createFavicon({
      sourceFile: joinPath(fixturesPath, 'nonSquare.svg'),
      outputDir: getTmpDir(),
      warn,
    })
    expect(warn).toHaveBeenCalledWith(
      'Source image is not square - it is HIGHLY recommended that input image is square'
    )
  })

  test('should not warn on non-square images if warnings are disabled', async () => {
    const spy = vi.spyOn(console, 'warn').mockImplementation(() => null)
    await createFavicon({
      sourceFile: joinPath(fixturesPath, 'nonSquare.svg'),
      warn: false,
      outputDir: getTmpDir(),
    })
    expect(spy).not.toHaveBeenCalled()
    spy.mockRestore()
  })

  test('should output square images from non-square input', async () => {
    const outputDir = getTmpDir()
    const sourceFile = joinPath(fixturesPath, 'nonSquare.svg')
    await createFavicon({
      sourceFile,
      warn: false,
      outputDir,
    })

    expect(await hashFile(joinPath(outputDir, 'favicon.svg'))).toBe(await hashFile(sourceFile))

    expect(await sharp(joinPath(outputDir, 'apple-touch-icon.png')).metadata()).toMatchObject({
      width: 180,
      height: 180,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-512.png')).metadata()).toMatchObject({
      width: 512,
      height: 512,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-192.png')).metadata()).toMatchObject({
      width: 192,
      height: 192,
      format: 'png',
      hasAlpha: true,
    })

    expect((await stat(joinPath(outputDir, 'favicon.ico'))).size).toBeGreaterThan(0)
    expect((await stat(joinPath(outputDir, 'favicon.ico'))).size).toBeGreaterThan(0)
  })

  test('should generate all variations from valid SVG', async () => {
    const outputDir = getTmpDir('valid')
    const sourceFile = joinPath(fixturesPath, 'mead.svg')
    const result = await createFavicon({
      sourceFile,
      outputDir,
    })

    expect(await hashFile(joinPath(outputDir, 'favicon.svg'))).toBe(await hashFile(sourceFile))

    expect(result.html).toMatchInlineSnapshot(`
      "<link rel=\\"icon\\" href=\\"/favicon.ico\\" sizes=\\"any\\">
      <link rel=\\"icon\\" href=\\"/favicon.svg\\" type=\\"image/svg+xml\\">
      <link rel=\\"apple-touch-icon\\" href=\\"/apple-touch-icon.png\\">
      <link rel=\\"manifest\\" href=\\"/manifest.webmanifest\\">"
    `)

    expect(await sharp(joinPath(outputDir, 'apple-touch-icon.png')).metadata()).toMatchObject({
      width: 180,
      height: 180,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-512.png')).metadata()).toMatchObject({
      width: 512,
      height: 512,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-192.png')).metadata()).toMatchObject({
      width: 192,
      height: 192,
      format: 'png',
      hasAlpha: true,
    })

    expect((await stat(joinPath(outputDir, 'favicon.ico'))).size).toBeGreaterThan(0)

    expect(JSON.parse(await readFile(joinPath(outputDir, 'manifest.webmanifest'), 'utf8')))
      .toMatchInlineSnapshot(`
      {
        "icons": [
          {
            "sizes": "192x192",
            "src": "/favicon-192.png",
            "type": "image/png",
          },
          {
            "sizes": "512x512",
            "src": "/favicon-512.png",
            "type": "image/png",
          },
        ],
      }
    `)
  })

  test('should skip writing webmanifest if `manifest` is set to `false`', async () => {
    const outputDir = getTmpDir('nomanifest')
    const sourceFile = joinPath(fixturesPath, 'mead.svg')
    const result = await createFavicon({
      sourceFile,
      outputDir,
      manifest: false,
    })

    expect(await hashFile(joinPath(outputDir, 'favicon.svg'))).toBe(await hashFile(sourceFile))
    expect(existsSync(joinPath(outputDir, 'manifest.webmanifest'))).toBe(false)

    expect(result.html).toMatchInlineSnapshot(`
      "<link rel=\\"icon\\" href=\\"/favicon.ico\\" sizes=\\"any\\">
      <link rel=\\"icon\\" href=\\"/favicon.svg\\" type=\\"image/svg+xml\\">
      <link rel=\\"apple-touch-icon\\" href=\\"/apple-touch-icon.png\\">"
    `)
  })

  test('should skip existing files by default', async () => {
    const warn = vi.fn()
    const outputDir = getTmpDir('existing')
    const sourceFile = joinPath(fixturesPath, 'mead.svg')

    // Place copies of _some_ file for all the below files, just to have something to
    // check whether or not gets overwritten later down
    const doNotReplaceFile = joinPath(__dirname, '..', 'README.md')
    const doNotReplaceHash = await hashFile(doNotReplaceFile)
    const targetFiles = [
      'favicon-512.png',
      'favicon-192.png',
      'apple-touch-icon.png',
      'favicon.ico',
      'manifest.webmanifest',
      'favicon.svg',
    ]

    await mkdir(outputDir, {recursive: true})
    await Promise.all(
      targetFiles.map((target) => copyFile(doNotReplaceFile, joinPath(outputDir, target)))
    )

    await createFavicon({
      sourceFile,
      outputDir,
      warn,
    })

    for (const target of targetFiles) {
      expect(await hashFile(joinPath(outputDir, target))).toBe(doNotReplaceHash)
      expect(warn).toHaveBeenCalledWith(`File ${target} already exists - skipping`)
    }
  })

  test('should replace existing files with `overwrite: true`', async () => {
    const warn = vi.fn()
    const outputDir = getTmpDir('overwrite')
    const sourceFile = joinPath(fixturesPath, 'mead.svg')

    // Place copies of _some_ file for all the below files, just to have something to
    // check whether or not gets overwritten later down
    const doNotReplaceFile = joinPath(__dirname, '..', 'README.md')
    const doNotReplaceHash = await hashFile(doNotReplaceFile)
    const targetFiles = [
      'favicon-512.png',
      'favicon-192.png',
      'apple-touch-icon.png',
      'favicon.ico',
      'manifest.webmanifest',
      'favicon.svg',
    ]

    await mkdir(outputDir, {recursive: true})
    await Promise.all(
      targetFiles.map((target) => copyFile(doNotReplaceFile, joinPath(outputDir, target)))
    )

    await createFavicon({
      overwrite: true,
      sourceFile,
      outputDir,
      warn,
    })

    for (const target of targetFiles) {
      expect(await hashFile(joinPath(outputDir, target))).not.toBe(doNotReplaceHash)
      expect(warn).toHaveBeenCalledWith(`File ${target} already exists - replacing`)
    }
  })

  test('should generate all variations from valid SVG (over HTTP)', async () => {
    const outputDir = getTmpDir('valid')
    const sourceFile = joinPath(fixturesPath, 'mead.svg')
    const result = await createFavicon({
      sourceFile: 'http://localhost:27344/mead.svg',
      outputDir,
    })

    expect(await hashFile(joinPath(outputDir, 'favicon.svg'))).toBe(await hashFile(sourceFile))

    expect(result.html).toMatchInlineSnapshot(`
      "<link rel=\\"icon\\" href=\\"/favicon.ico\\" sizes=\\"any\\">
      <link rel=\\"icon\\" href=\\"/favicon.svg\\" type=\\"image/svg+xml\\">
      <link rel=\\"apple-touch-icon\\" href=\\"/apple-touch-icon.png\\">
      <link rel=\\"manifest\\" href=\\"/manifest.webmanifest\\">"
    `)

    expect(await sharp(joinPath(outputDir, 'apple-touch-icon.png')).metadata()).toMatchObject({
      width: 180,
      height: 180,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-512.png')).metadata()).toMatchObject({
      width: 512,
      height: 512,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-192.png')).metadata()).toMatchObject({
      width: 192,
      height: 192,
      format: 'png',
      hasAlpha: true,
    })

    expect((await stat(joinPath(outputDir, 'favicon.ico'))).size).toBeGreaterThan(0)

    expect(JSON.parse(await readFile(joinPath(outputDir, 'manifest.webmanifest'), 'utf8')))
      .toMatchInlineSnapshot(`
      {
        "icons": [
          {
            "sizes": "192x192",
            "src": "/favicon-192.png",
            "type": "image/png",
          },
          {
            "sizes": "512x512",
            "src": "/favicon-512.png",
            "type": "image/png",
          },
        ],
      }
    `)
  })

  test('should generate all variations except SVG from PNG input, prints warning', async () => {
    const warn = vi.fn()
    const outputDir = getTmpDir('valid-png')
    const sourceFile = joinPath(fixturesPath, 'mead.png')
    const result = await createFavicon({
      sourceFile,
      outputDir,
      warn,
    })

    expect(warn).toHaveBeenCalledWith('Source image is not an SVG - skipping SVG output')

    // eslint-disable-next-line no-sync
    expect(existsSync(joinPath(outputDir, 'favicon.svg'))).toBe(false)

    expect(result.html).toMatchInlineSnapshot(`
      "<link rel=\\"icon\\" href=\\"/favicon.ico\\" sizes=\\"any\\">
      <link rel=\\"apple-touch-icon\\" href=\\"/apple-touch-icon.png\\">
      <link rel=\\"manifest\\" href=\\"/manifest.webmanifest\\">"
    `)

    expect(await sharp(joinPath(outputDir, 'apple-touch-icon.png')).metadata()).toMatchObject({
      width: 180,
      height: 180,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-512.png')).metadata()).toMatchObject({
      width: 512,
      height: 512,
      format: 'png',
      hasAlpha: true,
    })

    expect(await sharp(joinPath(outputDir, 'favicon-192.png')).metadata()).toMatchObject({
      width: 192,
      height: 192,
      format: 'png',
      hasAlpha: true,
    })

    expect((await stat(joinPath(outputDir, 'favicon.ico'))).size).toBeGreaterThan(0)

    expect(JSON.parse(await readFile(joinPath(outputDir, 'manifest.webmanifest'), 'utf8')))
      .toMatchInlineSnapshot(`
      {
        "icons": [
          {
            "sizes": "192x192",
            "src": "/favicon-192.png",
            "type": "image/png",
          },
          {
            "sizes": "512x512",
            "src": "/favicon-512.png",
            "type": "image/png",
          },
        ],
      }
    `)
  })

  test('should be able to specify custom base path', async () => {
    const outputDir = getTmpDir('custom-basepath')
    const result = await createFavicon({
      sourceFile: joinPath(fixturesPath, 'mead.svg'),
      outputDir,
      basePath: '/foo/bar/',
    })

    expect(result.html).toMatchInlineSnapshot(`
      "<link rel=\\"icon\\" href=\\"/foo/bar/favicon.ico\\" sizes=\\"any\\">
      <link rel=\\"icon\\" href=\\"/foo/bar/favicon.svg\\" type=\\"image/svg+xml\\">
      <link rel=\\"apple-touch-icon\\" href=\\"/foo/bar/apple-touch-icon.png\\">
      <link rel=\\"manifest\\" href=\\"/foo/bar/manifest.webmanifest\\">"
    `)

    expect(JSON.parse(await readFile(joinPath(outputDir, 'manifest.webmanifest'), 'utf8')))
      .toMatchInlineSnapshot(`
      {
        "icons": [
          {
            "sizes": "192x192",
            "src": "/foo/bar/favicon-192.png",
            "type": "image/png",
          },
          {
            "sizes": "512x512",
            "src": "/foo/bar/favicon-512.png",
            "type": "image/png",
          },
        ],
      }
    `)
  })
})


src/cli.ts
#!/usr/bin/env node
/* eslint-disable no-process-exit, no-console */
import cac from 'cac'
import {version} from '../package.json'
import {createFavicon} from './favicon.js'

const cli = cac('create-favicon')

cli
  .command('<source-file> [output-dir]', 'Generate favicons from a source image', {
    ignoreOptionDefaultValue: true,
  })
  .option('--overwrite', 'Overwrite existing files', {default: false})
  .option('--base-path <path>', 'Base path for printed HTML and web manifest', {default: '/'})
  .option('--no-manifest', 'Skip outputting a webmanifest', {default: false})
  .option('--no-warn', 'Disable warnings', {default: false})
  .example((name) => `${name} source.svg`)
  .example((name) => `${name} https://example.com/source.png --output-dir icons`)
  .action(async (sourceFile, outputDir, flags) => {
    const {basePath, warn, overwrite, manifest} = flags
    const options = {sourceFile, outputDir, basePath, warn, overwrite, manifest}

    try {
      const result = await createFavicon(options)
      console.log(result.html)
    } catch (err) {
      console.error(err)
      process.exit(1)
    }
  })

// Display help message when `-h` or `--help` appears
cli.help()

// Display version number when `-v` or `--version` appears
// It's also used in help message
cli.version(version)

// Trigger parsing and execute commands
try {
  cli.parse()
} catch (err: unknown) {
  if (!(err instanceof Error)) {
    throw new Error(`Unknown error: ${err}`)
  }

  if (err.message.includes('missing required args')) {
    cli.outputHelp()
    process.exit(1)
  }

  console.error(err)
  process.exit(1)
}


src/ico.ts
/**
 * Sharp instance to ICO buffer
 * Based on https://github.com/shtse8/sharp-to-ico
 * MIT (c) Weilin Shi
 */
import type {Sharp, OutputInfo} from 'sharp'

const BITMAP_SIZE = 40
const DIRECTORY_SIZE = 16
const HEADER_SIZE = 6

/**
 * Converts a sharp image instance to an ICO buffer, with a single 32x32 image
 *
 * @param img - The base image to use.
 * @returns A Buffer instance containing the ICO file
 * @internal
 */
export async function sharpToIco(img: Sharp): Promise<Buffer> {
  const {data, info} = await img
    .clone()
    .png()
    .resize(32, 32, {kernel: 'cubic'})
    .raw()
    .toBuffer({resolveWithObject: true})

  const header = getHeader()
  let offset = header.length + 16
  const dir = getDir(info, offset)
  const bmpInfoHeader = getBmpInfoHeader(info)
  const dib = getDib(data, info)

  const len = header.length + dir.length + bmpInfoHeader.length + dib.length
  offset += bmpInfoHeader.length + dib.length

  return Buffer.concat([header, dir, bmpInfoHeader, dib], len)
}

// https://en.wikipedia.org/wiki/ICO_(file_format)
function getHeader() {
  const buf = Buffer.alloc(HEADER_SIZE)

  buf.writeUInt16LE(0, 0) // Reserved. Must always be 0.
  buf.writeUInt16LE(1, 2) // Specifies image type: 1 for icon (.ICO) image
  buf.writeUInt16LE(1, 4) // Specifies number of images in the file.

  return buf
}

function getDir(info: OutputInfo, offset: number) {
  const buf = Buffer.alloc(DIRECTORY_SIZE)
  const size = info.size + BITMAP_SIZE

  buf.writeUInt8(info.width, 0) // Specifies image width in pixels.
  buf.writeUInt8(info.height, 1) // Specifies image height in pixels.
  buf.writeUInt8(0, 2) // Should be 0 if the image does not use a color palette.
  buf.writeUInt8(0, 3) // Reserved. Should be 0.
  buf.writeUInt16LE(1, 4) // Specifies color planes. Should be 0 or 1.
  buf.writeUInt16LE(32, 6) // Specifies bits per pixel.
  buf.writeUInt32LE(size, 8) // Specifies the size of the image's data in bytes
  buf.writeUInt32LE(offset, 12) // Specifies the offset of BMP or PNG data from the beginning of the ICO/CUR file

  return buf
}

// https://en.wikipedia.org/wiki/BMP_file_format
function getBmpInfoHeader(info: OutputInfo) {
  const buf = Buffer.alloc(BITMAP_SIZE)
  // https://en.wikipedia.org/wiki/ICO_(file_format)
  // ...Even if the AND mask is not supplied,
  // if the image is in Windows BMP format,
  // the BMP header must still specify a doubled height.
  const height = 64
  const width = 32
  const bpp = 32

  buf.writeUInt32LE(BITMAP_SIZE, 0) // The size of this header (40 bytes)
  buf.writeInt32LE(width, 4) // The bitmap width in pixels (signed integer)
  buf.writeInt32LE(height, 8) // The bitmap height in pixels (signed integer)
  buf.writeUInt16LE(1, 12) // The number of color planes (must be 1)
  buf.writeUInt16LE(bpp, 14) // The number of bits per pixel
  buf.writeUInt32LE(0, 16) // The compression method being used.
  buf.writeUInt32LE(info.size, 20) // The image size.
  buf.writeInt32LE(0, 24) // The horizontal resolution of the image. (signed integer)
  buf.writeInt32LE(0, 28) // The vertical resolution of the image. (signed integer)
  buf.writeUInt32LE(0, 32) // The number of colors in the color palette, or 0 to default to 2n
  buf.writeUInt32LE(0, 36) // The number of important colors used, or 0 when every color is important; generally ignored.

  return buf
}

// https://en.wikipedia.org/wiki/BMP_file_format
// Note that the bitmap data starts with the lower left hand corner of the image.
// blue green red alpha in order
function getDib(data: Buffer, info: OutputInfo) {
  const bpp = 4
  const cols = info.width * bpp
  const rows = info.height * cols
  const end = rows - cols
  const buf = Buffer.alloc(info.size)
  // xor map
  for (let row = 0; row < rows; row += cols) {
    for (let col = 0; col < cols; col += bpp) {
      let pos = row + col
      const r = data.readUInt8(pos)
      const g = data.readUInt8(pos + 1)
      const b = data.readUInt8(pos + 2)
      const a = data.readUInt8(pos + 3)

      pos = end - row + col
      buf.writeUInt8(b, pos)
      buf.writeUInt8(g, pos + 1)
      buf.writeUInt8(r, pos + 2)
      buf.writeUInt8(a, pos + 3)
    }
  }

  return buf
}


src/types.ts
/**
 * Options for the favicon generator
 *
 * @public
 */
export interface FaviconOptions {
  /**
   * The source image to use for the favicon. Can be a file path, a URL or a Buffer.
   *
   * Needs to be a square image with a minimum size of 512x512 pixels,
   * and should preferably be a SVG.
   */
  sourceFile: string | Buffer

  /**
   * Whether or not to overwrite existing files.
   *
   * Defaults to `false`.
   */
  overwrite?: boolean

  /**
   * Whether or not to output a web manifest file.
   *
   * Defaults to `true`.
   */
  manifest?: boolean

  /**
   * The output directory for the generated favicons.
   *
   * Defaults to a `favicons` folder within the current working directory.
   */
  outputDir?: string

  /**
   * Function to use for printing warnings.
   * Pass `false` to disable warnings.
   *
   * Defaults to `console.warn`.
   */
  warn?: typeof console.warn | false

  /**
   * Base path to use for printed HTML and web manifest files.
   *
   * Defaults to `/`.
   */
  basePath?: string
}

/**
 * The result of a favicon creation run
 *
 * @public
 */
export interface FaviconResult {
  /**
   * The HTML code to place in the `<head>` of your HTML document.
   * Note: You may need to adjust paths (use `basePath` option to programatically do so).
   */
  html: string
}


src/favicon.ts
import {join as joinPath, resolve as resolvePath} from 'node:path'
import {mkdir, stat, readFile, writeFile} from 'node:fs/promises'
import sharp from 'sharp'
import {request} from 'undici'
import {sharpToIco} from './ico.js'
import {generateWebManifest} from './webmanifest.js'
import type {FaviconOptions, FaviconResult} from './types.js'

/**
 * Create variations of a favicon from a source image
 *
 * @param options - Options for the favicon generator
 * @returns An object containing results of the favicon generation
 * @public
 */
export async function createFavicon(options: FaviconOptions): Promise<FaviconResult> {
  if (!options) {
    throw new Error('No options specified')
  }

  const {
    sourceFile,
    outputDir = joinPath(process.cwd(), 'favicons'),
    warn = console.warn,
    basePath = '/',
    overwrite = false,
    manifest = true,
  } = options

  if (!sourceFile) {
    throw new Error('No source file specified')
  }

  if (typeof sourceFile !== 'string' && !Buffer.isBuffer(sourceFile)) {
    throw new Error('Source file must be a string (file path or URL) or a Buffer')
  }

  const isBuffer = Buffer.isBuffer(sourceFile)
  const isUrl = !isBuffer && /^https?:\/\//.test(sourceFile)

  let source = Buffer.isBuffer(sourceFile) ? sourceFile : undefined
  if (isUrl) {
    source = await downloadImage(sourceFile)
  } else if (!isBuffer) {
    source = await readImage(resolvePath(process.cwd(), sourceFile))
  }

  if (!source) {
    throw new Error('Could not read source image')
  }

  const image = sharp(source)
  const {width, height, format} = await image.metadata()
  if (!width || !height) {
    throw new Error('Could not read image dimensions')
  }

  if (format !== 'svg' && (width < 512 || height < 512)) {
    throw new Error('Source image must be at least 512x512 pixels')
  }

  if (format !== 'svg') {
    printWarning('Source image is not an SVG - skipping SVG output', warn)
  }

  await mkdir(outputDir, {recursive: true})

  const base = image.ensureAlpha()

  if (width !== height) {
    printWarning(
      'Source image is not square - it is HIGHLY recommended that input image is square',
      warn
    )
    const size = Math.max(width, height, 512)
    base.resize(size, size, {fit: 'contain', background: 'transparent'})
  }

  async function maybeWriteFile(name: string, writer: (fullPath: string) => Promise<any>) {
    const fullPath = joinPath(outputDir, name)
    const exists = await fileExists(fullPath)

    if (exists && overwrite) {
      printWarning(`File ${name} already exists - replacing`, warn)
      return writer(fullPath)
    }

    if (exists && !overwrite) {
      return printWarning(`File ${name} already exists - skipping`, warn)
    }

    return writer(fullPath)
  }

  // 512x512 and 192x192 for Android devices
  await maybeWriteFile('favicon-512.png', (path) =>
    base.clone().resize(512, 512).png().toFile(path)
  )
  await maybeWriteFile('favicon-192.png', (path) =>
    base.clone().resize(192, 192).png().toFile(path)
  )

  // 180x180 for iOS devices
  await maybeWriteFile('apple-touch-icon.png', (path) =>
    base.clone().resize(180, 180).png().toFile(path)
  )

  // 32x32 favicon for older browsers
  await maybeWriteFile('favicon.ico', async (path) =>
    writeFile(path, await sharpToIco(base.clone()))
  )

  // Web manifest file pointing to the generated files
  if (manifest) {
    await maybeWriteFile('manifest.webmanifest', (path) =>
      writeFile(path, generateWebManifest(basePath))
    )
  }

  // If the input is an SVG, pass-through the original SVG as well
  if (format === 'svg' && source) {
    const inputSource = source
    await maybeWriteFile('favicon.svg', (path) => writeFile(path, inputSource))
  }

  // Generate the HTML needed for the `<head>` of the HTML document
  const html = generateHtml({basePath, hasSvg: format === 'svg', manifest})

  return {html}
}

/**
 * Generates the HTML required for the `<head>` of the HTML document
 *
 * @param options - Options for the HTML generator
 * @returns The `<link>` tags required, unindented and separated by newlines
 * @internal
 */
function generateHtml(options: {basePath: string; hasSvg: boolean; manifest: boolean}): string {
  const {basePath, hasSvg, manifest} = options
  const base = basePath.endsWith('/') ? basePath.replace(/\/+$/, '') : basePath

  const links = [`<link rel="icon" href="${base}/favicon.ico" sizes="any">`]

  if (hasSvg) {
    links.push(`<link rel="icon" href="${base}/favicon.svg" type="image/svg+xml">`)
  }

  links.push(`<link rel="apple-touch-icon" href="${base}/apple-touch-icon.png">`)

  if (manifest) {
    links.push(`<link rel="manifest" href="${base}/manifest.webmanifest">`)
  }

  return links.join('\n')
}

/**
 * Downloads an image from a URL to a Buffer
 *
 * @param url - The URL to download
 * @returns A buffer containing the downloaded image
 * @internal
 */
async function downloadImage(url: string): Promise<Buffer> {
  const {statusCode, body} = await request(url, {reset: true}).catch((err: unknown) => {
    const message = err instanceof Error ? err.message : `${err}`
    throw new Error(`Failed fetching image from "${url}": ${message}`)
  })

  if (statusCode !== 200) {
    throw new Error(`Failed fetching image from "${url}": Server returned HTTP ${statusCode}`)
  }

  const buffers: Buffer[] = []
  for await (const chunk of body) {
    buffers.push(chunk)
  }

  return Buffer.concat(buffers)
}

/**
 * Read an image from the filesystem to a buffer, with a helpful error message on errors
 *
 * @param filePath - The path to the file to read
 * @returns A buffer containing the image data
 * @internal
 */
async function readImage(filePath: string): Promise<Buffer> {
  try {
    return await readFile(filePath)
  } catch (err) {
    const message = err instanceof Error ? err.message : `${err}`
    throw new Error(`Could not read file "${filePath}": ${message}`)
  }
}

/**
 * Print a warning message with the given warner function.
 * If the warner is the default `console.warn`, we will add a prefix and color it yellow.
 *
 * @param message - Message to print
 * @param warner - Function to use for printing the message
 * @internal
 */
function printWarning(message: string, warner: typeof console.warn | false): void {
  if (!warner) {
    return
  }

  warner(warner === console.warn ? `\u001b[93m[warn]\u001b[39m ${message}` : message)
}

/**
 * Checks if a file exists or not, with a friendly error if the existing path is not a file
 *
 * @param filePath - File path to check
 * @returns Promise resolving to true if it exists, false otherwise
 */
async function fileExists(filePath: string): Promise<boolean> {
  try {
    const file = await stat(filePath)
    if (!file.isFile()) {
      throw new Error(`"${filePath}" is not a file`)
    }
    return true
  } catch (err: unknown) {
    if (!(err instanceof Error)) {
      throw new Error(`${err}`)
    }

    if ('code' in err && err.code === 'ENOENT') {
      return false
    }

    throw err
  }
}
